src/cargo/sources/git/utils.rs

```diff
@@ -355,12 +364,52 @@ impl<'a> GitCheckout<'a> {
                     "Skipping",
                     format!(
                         "git submodule `{}` due to update strategy in .gitmodules",
-                        url
+                        child_url_str
                     ),
                 )?;
                 return Ok(());
             }
 
+            // There are a few possible cases here:
+            // 1. Submodule URL is not relative.
+            //     Happy path, Url is just the submodule url.
+            // 2. Submodule URL is relative and does specify ssh/https/file/etc.
+            //     We combine the parent path and relative path.
+            // 3. Submodule URL is relative and does not specify ssh/https/file/etc.
+            //     In which case we fail to parse with the error ParseError::RelativeUrlWithoutBase.
+            //     We then combine the relative url with the host/protocol from the parent url.
+            // 4. We fail to parse submodule url for some reason.
+            //     Error. Gets passed up.
+
+            let join_urls = || {
+                let path = parent_remote_url.path();
+                let mut new_parent_remote_url = parent_remote_url.clone();
+                new_parent_remote_url.set_path(&format!("{}/", path));
```

Comment:
The only thing I'd suggest is to check whether parent path already ends with '/'. `Url` docs state that multiple '/'s are important.